clc/store/store.go

```diff
@@ -0,0 +1,189 @@
+)
+
+type StoreAccessor struct {
+	store *Store
+	mu    sync.Mutex
+}
+
+func NewStoreAccessor(dir string, logger log.Logger) *StoreAccessor {
+	return &StoreAccessor{
+		store: newStore(dir, logger),
+	}
+}
+
+func (sa *StoreAccessor) WithLock(fn func(s *Store) (any, error)) (any, error) {
+	sa.mu.Lock()
+	defer sa.mu.Unlock()
+	if err := sa.store.open(); err != nil {
+		return nil, err
+	}
+	defer sa.store.close()
+	return fn(sa.store)
+}
+
+type Store struct {
+	dir    string
+	db     *badger.DB
+	logger log.Logger
+}
+
+func newStore(dir string, logger log.Logger) *Store {
+	return &Store{
+		dir:    dir,
+		logger: logger,
+	}
+}
+
+func (s *Store) open() error {
+	opts := badger.DefaultOptions(s.dir)
+	opts.Logger = badgerLogger{s.logger}
+	db, err := badger.Open(opts)
+	if err != nil {
+		return err
+	}
+	s.db = db
+	return nil
+}
+
+type badgerLogger struct {
+	logger log.Logger
+}
+
+func (l badgerLogger) Errorf(log string, args ...any) {
+	l.logger.Debugf(log, args...)
+}
+func (l badgerLogger) Warningf(log string, args ...any) {
+	l.logger.Debugf(log, args...)
+}
+func (l badgerLogger) Infof(log string, args ...any) {
+	l.logger.Debugf(log, args...)
+}
+func (l badgerLogger) Debugf(log string, args ...any) {
+	l.logger.Debugf(log, args...)
+}
+
+func (s *Store) close() error {
+	return s.db.Close()
+}
+
+func (s *Store) SetEntry(key, val []byte) error {
+	err := s.db.Update(func(txn *badger.Txn) error {
+		e := badger.NewEntry(key, val)
+		return txn.SetEntry(e)
+	})
+	return err
+}
+
+func (s *Store) GetEntry(key []byte) ([]byte, error) {
+	var item []byte
+	err := s.db.View(func(txn *badger.Txn) error {
+		it, err := txn.Get(key)
+		if err != nil {
+			if errors.Is(err, badger.ErrKeyNotFound) {
+				return ErrKeyNotFound
+			}
+			return err
+		}
+		item, err = it.ValueCopy(nil)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+	return item, nil
+}
+
+func (s *Store) GetKeysWithPrefix(prefix string) ([][]byte, error) {
+	keys := [][]byte{}
+	pb := []byte(prefix)
+	opts := badger.DefaultIteratorOptions
+	opts.PrefetchValues = false
+	err := s.db.View(func(txn *badger.Txn) error {
+		it := txn.NewIterator(opts)
+		defer it.Close()
+		for it.Seek(pb); it.ValidForPrefix(pb); it.Next() {
+			keyCopy := it.Item().KeyCopy(nil)
+			keys = append(keys, keyCopy)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return keys, nil
+}
+
+type updateFn = func(current []byte, found bool) []byte
+
+func (s *Store) UpdateEntry(key []byte, f updateFn) error {
+	var item []byte
+	err := s.db.Update(func(txn *badger.Txn) error {
+		it, err := txn.Get(key)
+		found := true
+		if err != nil {
+			if errors.Is(err, badger.ErrKeyNotFound) {
+				found = false
+			} else {
+				return err
+			}
+		} else {
+			item, err = it.ValueCopy(nil)
+			if err != nil {
+				return err
+			}
+		}
+		newVal := f(item, found)
+		return txn.Set(key, newVal)
+	})
+	return err
+}
+
+type foreachFn = func(key, val []byte)
```

Comment:
```suggestion
type foreachFunc = func(key, val []byte)
```

Reply:
How about:
```
type ForeachFunc = func(key, val []byte) (ok bool, err error)
```
So if the functio returns `ok == false` we can exit the loop early.
And if there's an error we can return it from `RunForeachWithPrefix`.

Reply:
Alright, I will change the signature